File: src/weaponsprocurement/gui/WimGuiModalListRenderer.java
package weaponsprocurement.gui;

import com.fs.starfarer.api.ui.CustomPanelAPI;

import java.util.List;

final class WimGuiModalListRenderer {
    interface ScrollRowFactory<A> {
        WimGuiListRow<A> createScrollRow(String label, int scrollDelta);
    }

    interface ExtraGapProvider<A> {
        float extraGapBefore(WimGuiListRow<A> row);
    }

    private WimGuiModalListRenderer() {
    }

    static <A> WimGuiModalListRenderResult render(CustomPanelAPI root,
                                                  List<WimGuiListRow<A>> rows,
                                                  int requestedOffset,
                                                  WimGuiModalListSpec spec,
                                                  ScrollRowFactory<A> scrollRowFactory,
                                                  ExtraGapProvider<A> extraGapProvider,
                                                  List<WimGuiButtonBinding<A>> buttons) {
        WimGuiModalListLayout<WimGuiListRow<A>> layout = WimGuiModalListLayout.compute(
                rows,
                requestedOffset,
                spec.panelTop,
                spec.panelHeight,
                spec.panelWidth,
                spec.modal,
                gapProvider(extraGapProvider));
        WimGuiScrollSlice<WimGuiListRow<A>> slice = layout.scrollSlice;
        CustomPanelAPI listPanel = root.createCustomPanel(
                spec.panelWidth,
                layout.panelHeight,
                new WimGuiPanelPlugin(spec.panelFill, spec.panelBorder));
        root.addComponent(listPanel).inTL(spec.panelLeft, layout.panelTop);

        float y = spec.rowHorizontalPad + finalPageTopOffset(slice, layout.panelHeight, spec, extraGapProvider);
        if (slice.hasAbove) {
            renderRow(listPanel, scrollRowFactory.createScrollRow(WimGuiScrollIndicator.ABOVE,
                    -layout.pageDelta()), y, spec, buttons);
            y += spec.rowHeight + spec.rowGap;
        }
        for (int i = 0; i < slice.items.size(); i++) {
            WimGuiListRow<A> row = slice.items.get(i);
            y += extraGapProvider == null ? 0f : extraGapProvider.extraGapBefore(row);
            renderRow(listPanel, row, y, spec, buttons);
            y += spec.rowHeight + spec.rowGap;
        }
        if (slice.hasBelow) {
            renderRow(listPanel, scrollRowFactory.createScrollRow(WimGuiScrollIndicator.BELOW,
                    layout.pageDelta()), y, spec, buttons);
        }
        return new WimGuiModalListRenderResult(
                new WimGuiListBounds(
                        layout.maxOffset,
                        spec.panelLeft,
                        layout.panelTop,
                        spec.panelWidth,
                        layout.panelHeight),
                slice.offset);
    }

    static <A> WimGuiListBounds renderAndStoreOffset(CustomPanelAPI root,
                                                     List<WimGuiListRow<A>> rows,
                                                     WimGuiScrollableListState state,
                                                     WimGuiModalListSpec spec,
                                                     ScrollRowFactory<A> scrollRowFactory,
                                                     ExtraGapProvider<A> extraGapProvider,
                                                     List<WimGuiButtonBinding<A>> buttons) {
        WimGuiModalListRenderResult result = render(
                root,
                rows,
                state == null ? 0 : state.getListScrollOffset(),
                spec,
                scrollRowFactory,
                extraGapProvider,
                buttons);
        if (state != null) {
            state.setListScrollOffset(result.getOffset());
        }
        return result.getBounds();
    }

    private static <A> void renderRow(CustomPanelAPI listPanel,
                                      WimGuiListRow<A> row,
                                      float y,
                                      WimGuiModalListSpec spec,
                                      List<WimGuiButtonBinding<A>> buttons) {
        WimGuiListRowRenderer.renderRow(
                listPanel,
                row,
                y,
                spec.rowHeight,
                spec.actionHeight,
                spec.rowHorizontalPad,
                spec.buttonGap,
                spec.textLeftPad,
                spec.minLabelWidth,
                spec.rowBorder,
                buttons);
    }

    private static <A> float finalPageTopOffset(WimGuiScrollSlice<WimGuiListRow<A>> slice,
                                                float panelHeight,
                                                WimGuiModalListSpec spec,
                                                ExtraGapProvider<A> extraGapProvider) {
        if (slice == null || !slice.hasAbove || slice.hasBelow || slice.items.isEmpty()) {
            return 0f;
        }
        float contentBottom = spec.rowHeight;
        for (int i = 0; i < slice.items.size(); i++) {
            WimGuiListRow<A> row = slice.items.get(i);
            contentBottom += spec.rowGap + gapBefore(row, extraGapProvider) + spec.rowHeight;
        }
        float innerHeight = panelHeight - 2f * spec.rowHorizontalPad;
        return Math.max(0f, innerHeight - contentBottom);
    }

    private static <A> float gapBefore(WimGuiListRow<A> row, ExtraGapProvider<A> extraGapProvider) {
        return extraGapProvider == null ? 0f : Math.max(0f, extraGapProvider.extraGapBefore(row));
    }

    private static <A> WimGuiScroll.ExtraGapProvider<WimGuiListRow<A>> gapProvider(
            final ExtraGapProvider<A> extraGapProvider) {
        if (extraGapProvider == null) {
            return null;
        }
        return new WimGuiScroll.ExtraGapProvider<WimGuiListRow<A>>() {
            @Override
            public float extraGapBefore(WimGuiListRow<A> row) {
                return extraGapProvider.extraGapBefore(row);
            }
        };
    }

}

File: src/main/kotlin/weaponsprocurement/gui/WimGuiModalListRenderer.kt
package weaponsprocurement.gui

import com.fs.starfarer.api.ui.CustomPanelAPI

class WimGuiModalListRenderer private constructor() {
    fun interface ScrollRowFactory<A> {
        fun createScrollRow(label: String, scrollDelta: Int): WimGuiListRow<A>
    }

    fun interface ExtraGapProvider<A> {
        fun extraGapBefore(row: WimGuiListRow<A>): Float
    }

    companion object {
        @JvmStatic
        fun <A> render(
            root: CustomPanelAPI,
            rows: List<WimGuiListRow<A>>,
            requestedOffset: Int,
            spec: WimGuiModalListSpec,
            scrollRowFactory: ScrollRowFactory<A>,
            extraGapProvider: ExtraGapProvider<A>?,
            buttons: MutableList<WimGuiButtonBinding<A>>,
        ): WimGuiModalListRenderResult {
            val layout = WimGuiModalListLayout.compute(
                rows,
                requestedOffset,
                spec.panelTop,
                spec.panelHeight,
                spec.panelWidth,
                spec.modal,
                gapProvider(extraGapProvider),
            )
            val slice = layout.scrollSlice
            val listPanel = root.createCustomPanel(
                spec.panelWidth,
                layout.panelHeight,
                WimGuiPanelPlugin(spec.panelFill, spec.panelBorder),
            )
            root.addComponent(listPanel).inTL(spec.panelLeft, layout.panelTop)

            var y = spec.rowHorizontalPad + finalPageTopOffset(slice, layout.panelHeight, spec, extraGapProvider)
            if (slice.hasAbove) {
                renderRow(listPanel, scrollRowFactory.createScrollRow(WimGuiScrollIndicator.ABOVE, -layout.pageDelta()), y, spec, buttons)
                y += spec.rowHeight + spec.rowGap
            }
            for (row in slice.items) {
                y += gapBefore(row, extraGapProvider)
                renderRow(listPanel, row, y, spec, buttons)
                y += spec.rowHeight + spec.rowGap
            }
            if (slice.hasBelow) {
                renderRow(listPanel, scrollRowFactory.createScrollRow(WimGuiScrollIndicator.BELOW, layout.pageDelta()), y, spec, buttons)
            }
            return WimGuiModalListRenderResult(
                WimGuiListBounds(layout.maxOffset, spec.panelLeft, layout.panelTop, spec.panelWidth, layout.panelHeight),
                slice.offset,
            )
        }

        @JvmStatic
        fun <A> renderAndStoreOffset(
            root: CustomPanelAPI,
            rows: List<WimGuiListRow<A>>,
            state: WimGuiScrollableListState?,
            spec: WimGuiModalListSpec,
            scrollRowFactory: ScrollRowFactory<A>,
            extraGapProvider: ExtraGapProvider<A>?,
            buttons: MutableList<WimGuiButtonBinding<A>>,
        ): WimGuiListBounds {
            val result = render(root, rows, state?.getListScrollOffset() ?: 0, spec, scrollRowFactory, extraGapProvider, buttons)
            state?.setListScrollOffset(result.offset)
            return result.bounds
        }

        private fun <A> renderRow(
            listPanel: CustomPanelAPI,
            row: WimGuiListRow<A>,
            y: Float,
            spec: WimGuiModalListSpec,
            buttons: MutableList<WimGuiButtonBinding<A>>,
        ) {
            WimGuiListRowRenderer.renderRow(
                listPanel,
                row,
                y,
                spec.rowHeight,
                spec.actionHeight,
                spec.rowHorizontalPad,
                spec.buttonGap,
                spec.textLeftPad,
                spec.minLabelWidth,
                spec.rowBorder,
                buttons,
            )
        }

        private fun <A> finalPageTopOffset(
            slice: WimGuiScrollSlice<WimGuiListRow<A>>?,
            panelHeight: Float,
            spec: WimGuiModalListSpec,
            extraGapProvider: ExtraGapProvider<A>?,
        ): Float {
            if (slice == null || !slice.hasAbove || slice.hasBelow || slice.items.isEmpty()) {
                return 0f
            }
            var contentBottom = spec.rowHeight
            for (row in slice.items) {
                contentBottom += spec.rowGap + gapBefore(row, extraGapProvider) + spec.rowHeight
            }
            val innerHeight = panelHeight - 2f * spec.rowHorizontalPad
            return Math.max(0f, innerHeight - contentBottom)
        }

        private fun <A> gapBefore(row: WimGuiListRow<A>, extraGapProvider: ExtraGapProvider<A>?): Float =
            if (extraGapProvider == null) 0f else Math.max(0f, extraGapProvider.extraGapBefore(row))

        private fun <A> gapProvider(extraGapProvider: ExtraGapProvider<A>?): WimGuiScroll.ExtraGapProvider<WimGuiListRow<A>>? {
            if (extraGapProvider == null) {
                return null
            }
            return WimGuiScroll.ExtraGapProvider { row -> extraGapProvider.extraGapBefore(row) }
        }
    }
}
